libs/utils/trace_analysis.py

```diff
@@ -950,3 +950,158 @@ def plotCPUFrequencyResidency(self, cpus=None):
                           labels,
                           loc='lower right')
             plt.subplots_adjust(hspace=0.5)
+
+    def getClusterFrequencyResidency(self, cluster):
+        """
+        Get a DataFrame with per cluster frequency residency, i.e. amount of
+        time spent at a given frequency in each cluster.
+
+        :param cluster: list of CPU IDs belonging to the cluster
+        :type cluster: list(int)
+        """
+
+        if not self.trace.hasEvents('cpu_frequency'):
+            logging.warn('Events [cpu_frequency] not found, '\
+                         'plot DISABLED!')
+            return None
+        if not self.trace.hasEvents('cpu_idle'):
+            logging.warn('Events [cpu_idle] not found, '\
+                         'plot DISABLED!')
+            return None
+
+        freq_dfr = self.trace.df('cpu_frequency')
+        cluster_freqs = freq_dfr[freq_dfr['cpu'].isin(cluster)]
+        idle_dfr = self.trace.df('cpu_idle')
+
+        ### Computer TOTAL Time ###
+        time_intervals = cluster_freqs.index[1:] - cluster_freqs.index[:-1]
+        total_time = pd.DataFrame({
+            'TOTAL Time' : time_intervals,
+            'Frequency [MHz]' : [f/1000 for f in cluster_freqs.iloc[:-1].frequency]
+        })
+        total_time = total_time.groupby(['Frequency [MHz]']).sum()
+
+        ### Compute ACTIVE Time ###
+        cpu_activation = {}
+        for cpu in cluster:
+            cpu_freqs = freq_dfr[freq_dfr['cpu'] == cpu]
+            cpu_idle = idle_dfr[idle_dfr['cpu_id'] == cpu]
+
+            # time instants where CPU goes to an idle state
+            idle_times = cpu_idle[cpu_idle['state'] != IDLE_STATE]
+            # time instants where CPU leaves an idle state
+            non_idle_times = cpu_idle[cpu_idle['state'] == IDLE_STATE]
+
+            if idle_times.index[0] < non_idle_times.index[0]:
+                idle_times = idle_times.iloc[1:]
+
+            activation_indexes = sorted(
+                non_idle_times.index.values.tolist() + \
+                idle_times.index.values.tolist()
+            )
+            if activation_indexes[0] != 0.0:
+                activation_indexes.insert(0, 0.0)
+                activations = [i%2 for i in range(len(activation_indexes))]
+            else:
+                activations = [(i+1)%2 for i in range(len(activation_indexes))]
+            cpu_activation[cpu] = pd.Series(activations,
+                                            index=activation_indexes)
+
+        activations = pd.DataFrame(cpu_activation)
+        activations.loc[0.0].fillna(0, inplace=True)
+        activations.fillna(method="ffill", inplace=True)
+
+        cluster_active = activations[activations.columns[0]].astype(int)
+        for c in range(1, len(activations.columns)):
+            cluster_active |= activations[activations.columns[c]].astype(int)
```

Comment:
You could use `reduce` instead of those 3 lines:
``` python
cluster_active = reduce(operator.or_, [cpu_activations.astype(int) for _, cpu_activations in activations.iteritems()])
```

You'd have to `import operator` at the beginning of the file.  Not sure whether the code is clearer.